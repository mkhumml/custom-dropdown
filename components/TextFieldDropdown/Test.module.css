@keyframes first {
  0% {
    transform: translateY(0);
  }
  100% {
    transform: translateY(0);
  }
}

@keyframes nth {
  0% {
    transform: translateY(-25px);
  }
  100% {
    transform: translateY(0);
  }
}

.Animated:first-child {
  animation-name: first;
  animation-duration: 0.75s;
}

.Animated {
  animation-name: nth;
  animation-duration: 0.75s;
}
