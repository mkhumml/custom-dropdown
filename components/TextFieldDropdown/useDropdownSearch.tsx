import { useEffect, useState } from "react";
import axios from "axios";

interface IUseDropdownSearchProps {
  state: string | null;
}

interface IUseDropdownSearchStateProps {
  state: string;
}

const useDropdownSearch = ({ state }: IUseDropdownSearchProps) => {
  const [data, setData] = useState<IUseDropdownSearchStateProps[] | null>(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    (async function () {
      try {
        setLoading(true);
        const response: any = await axios.get(
          "https://jsonplaceholder.typicode.com/users"
        );
        const responseData: IUseDropdownSearchStateProps[] = response.data;

        if (responseData) {
          if (state !== null && state !== "") {
            setData(
              responseData.filter((item: any) => {
                if (item.name.toLowerCase() === state.toLowerCase()) {
                  return null;
                } else {
                  return item.name.toLowerCase().includes(state.toLocaleLowerCase());
                }
              })
            );
          } else {
            setData(null);
          }
        }
      } catch (err: any) {
        setError(err);
      } finally {
        setLoading(false);
      }
    })();
  }, [state]);

  return { data, error, loading };
};

export default useDropdownSearch;
