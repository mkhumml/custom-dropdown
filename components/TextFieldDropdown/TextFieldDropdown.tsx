import React, { useEffect, useState } from "react";
import styles from "./Test.module.css";
import useDropdownSearch from "./useDropdownSearch";

interface ITextFieldDropdown {
  setDropdownValue: Function | undefined;
  dropdownValue: string | undefined;
  label: string;
  name: string;
}

const TextFieldDropdown = ({ setDropdownValue, label, name }: ITextFieldDropdown) => {
  const [state, setState] = useState("");
  const { data } = useDropdownSearch({ state });

  useEffect(() => {
    if (setDropdownValue) {
      setDropdownValue(state);
    }
  }, [state, setDropdownValue]);

  return (
    <div className="block p-4">
      <input
        value={state}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => setState(e.target.value)}
        id={name}
        type="text"
        autoComplete="off"
        className="shadow-inputBorder py-2 px-4 w-full focus:outline-0 focus:shadow-inputBorderFocus"
      />
      <div className="relative z-10">
        <span className="font-semibold absolute -top-[3.35rem] left-2 pl-1 pr-1 bg-white z-10">
          {label}
        </span>
        <div className="absolute z-10 w-full">
          {data?.map((item: any, key: number) => {
            return (
              <div
                onClick={(event: any) => {
                  setState(event.target.innerHTML);
                }}
                className={
                  "hover:bg-gray-100 hover:cursor-pointer hover:shadow-inputBorderFocus hover:relative hover:z-10 bg-white shadow-inputBorder py-2 px-4 " +
                  `${styles.Animated}`
                }
                key={key}
              >
                {item.name}
              </div>
            );
          })}
        </div>
      </div>
    </div>
  );
};

export default TextFieldDropdown;
