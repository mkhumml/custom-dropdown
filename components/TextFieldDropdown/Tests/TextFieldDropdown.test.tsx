import TextFieldDropdown from "../TextFieldDropdown";
import renderer from "react-test-renderer";

let mockData: any = null;

jest.mock("../useDropdownSearch", () => {
  return jest.fn(() => ({
    data: mockData,
    error: null,
    loading: false,
  }));
});


describe("TextFieldDropdown without Data", () => {
  test("Controlled Snapshot", () => {
    const tree = renderer.create(
      <TextFieldDropdown
        setDropdownValue={jest.fn((x) => x)}
        dropdownValue={""}
        label="DropdownControlled"
        name="DropdownControlledTest"
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
