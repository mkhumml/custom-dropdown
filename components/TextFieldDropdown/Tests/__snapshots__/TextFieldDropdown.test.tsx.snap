// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFieldDropdown without Data Controlled Snapshot 1`] = `
<div
  className="block p-4 border border-red-600"
>
  <input
    autoComplete="off"
    className="shadow-inputBorder py-2 px-4 w-full "
    id="DropdownControlledTest"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="relative"
  >
    <span
      className="font-semibold absolute -top-[3.35rem] left-2 pl-1 pr-1 bg-white z-inputLabel"
    >
      DropdownControlled
    </span>
    <div
      className="absolute z-dropdownOptionsContainer w-full  hidden"
    />
  </div>
</div>
`;
